File: BFS_DFS_etc_Data_structure/bfs_dfs.c
#include <stdio.h>
#include <stdlib.h>
// 이진 트리 노드 정의
typedef struct TreeNode{
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
} TreeNode;

TreeNode* createNode(int val){
    TreeNode* new_node = (TreeNode*)malloc(sizeof(TreeNode));
    new_node -> val = val;
    new_node -> left = NULL;
    new_node -> right = NULL;
    return new_node;
}

// BFS (Level Order Traversal)
void bfs(TreeNode* root) {
    if(!root) return;
    int front = 0, rear = 0;
    TreeNode* q[100]; // 임의 설정
    q[rear++] = root;
    while(front < rear){
        TreeNode* temp = q[front++];
        printf("%d ",temp->val);
        if(temp->left) q[rear++] = temp->left;
        if(temp->right) q[rear++] = temp->right;
    }
}

// DFS (Preorder Traversal)
void dfs(TreeNode* root) {
    if(!root) return;
    int idx = 0;
    TreeNode* s[100];
    s[idx++] = root;
    while(idx>0){
        TreeNode* temp = s[--idx];
        printf("%d ",temp->val);
        if(temp->right) s[idx++] = temp->right;
        if(temp->left) s[idx++] = temp->left;
    }
}

// 트리 생성 및 테스트용 main 함수
int main() {
    TreeNode* root = createNode(1);
    root->left = createNode(2);
    root->right = createNode(3);
    root->left->left = createNode(4);
    root->left->right = createNode(5);
    root->right->left = createNode(6);
    root->right->right = createNode(7);

    printf("BFS (Level Order Traversal):\n");
    bfs(root);
    printf("\nDFS (Preorder Traversal):\n");
    dfs(root);
    
    return 0;
}

File: BFS_DFS_etc_Data_structure/bfs_dfs.py
from collections import deque

# 이진 트리 노드 정의
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# BFS (Level Order Traversal)
def bfs(root:TreeNode)->None : # BFS는 QUEUE
    q = deque([root])
    while q:
        temp = q.popleft()
        print(temp.val,end=" ")
        if temp.left : q.append(temp.left)
        if temp.right : q.append(temp.right)

# DFS (Preorder Traversal)
def dfs(root:TreeNode) -> None:
    s = deque([root])
    while s:
        temp = s.pop()
        print(temp.val,end=" ")
        if temp.right : s.append(temp.right)
        if temp.left : s.append(temp.left)

# Python 실행파일
if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(7)

    print("BFS (Level Order Traversal):")
    bfs(root)
    print("\nDFS (Preorder Traversal):")
    dfs(root)